packages/blockchain/src/consensus/clique.ts

```diff
@@ -400,12 +418,17 @@ export class CliqueConsensus implements Consensus {
   /**
    * Returns a list with the current block signers
    */
-  cliqueActiveSigners(): Address[] {
+  cliqueActiveSigners(blockNum: bigint): Address[] {
     const signers = this._cliqueLatestSignerStates
     if (signers.length === 0) {
       return []
     }
-    return [...signers[signers.length - 1][1]]
+    for (let i = signers.length - 1; i >= 0; i--) {
+      if (signers[i][0] < blockNum) {
```

Comment:
Just clarify my understanding here, we're just returning the most recent signer state, (which might be from a block older than the current chain head if no signer state change occurs on the current block, right?)

Reply:
I think you got the idea here, but just for clarity:

- This solves the reorg problem, because before it always returned whatever active signers there were at current chain head. (On client this could result in execution vs block import being out of sync which thus require different active signers if you verify them)
- Note that `_cliqueLatestSignerStates` is sorted
- We return the first candidate which has active signers < requested block num (so the state we want)
- Note that signers always has 1 member